File: sdks/typescript/src/sdk/logs.ts
import { toHex } from "../crypto/identifiers.ts";
import {
  type CanonicalLogUrl,
  assertCanonicalLogUrl,
} from "../log/canonical-url.ts";
import { MockTransparencyLog } from "../log/mock-log.ts";
import {
  type TransparencyLogEntry,
  type TransparencyLogQueryResult,
} from "../log/types.ts";
import { base64urlEncode, decodeBase64url } from "./keys.ts";

export type MaybePromise<T> = T | Promise<T>;

export type SdkSubmissionLog = {
  logUrl: CanonicalLogUrl;
  append(
    envelope: Uint8Array,
    integratedTime?: string,
  ): MaybePromise<TransparencyLogEntry>;
};

export type SelloHttpLogOptions = {
  endpoint?: string;
  logUrl?: string;
  headers?: Record<string, string>;
  fetch?: typeof fetch;
};

export type SerializedTransparencyLogEntry = {
  logUrl: string;
  index: number;
  integratedTime: string;
  envelope: string;
  proof: unknown;
};

export function memory(url: string): MockTransparencyLog {
  return new MockTransparencyLog(toCanonicalLogUrl(url));
}

export function http(url: string, options: SelloHttpLogOptions = {}): HttpSelloLog {
  return new HttpSelloLog(url, options);
}

export async function queryHttpLogByTokenRef(
  input: {
    endpoint: string;
    tokenRef: Uint8Array;
    headers?: Record<string, string>;
    fetch?: typeof fetch;
  },
): Promise<TransparencyLogQueryResult> {
  const fetcher = input.fetch ?? fetch;
  const url = buildUrl(input.endpoint, `/entries?sello_token_ref=${toHex(input.tokenRef)}`);
  const response = await fetcher(url, {
    method: "GET",
    headers: input.headers,
  });

  if (!response.ok) {
    throw new TypeError(`Sello log query failed with HTTP ${response.status}`);
  }

  const decoded = await response.json();
  if (!isRecord(decoded) || !Array.isArray(decoded.entries)) {
    throw new TypeError("Sello log query response must contain entries");
  }

  const completeness =
    decoded.completeness === "complete" ? "complete" : "discovery-only";

  return {
    completeness,
    entries: decoded.entries.map(deserializeEntry),
  };
}

export class HttpSelloLog {
  readonly logUrl: CanonicalLogUrl;
  readonly endpoint: string;
  readonly #headers?: Record<string, string>;
  readonly #fetch: typeof fetch;

  constructor(url: string, options: SelloHttpLogOptions = {}) {
    this.logUrl = toCanonicalLogUrl(options.logUrl ?? url);
    this.endpoint = normalizeEndpoint(options.endpoint ?? url);
    this.#headers = options.headers;
    this.#fetch = options.fetch ?? fetch;
  }

  async append(
    envelope: Uint8Array,
    integratedTime?: string,
  ): Promise<TransparencyLogEntry> {
    const response = await this.#fetch(buildUrl(this.endpoint, "/entries"), {
      method: "POST",
      headers: {
        "content-type": "application/json",
        ...this.#headers,
      },
      body: JSON.stringify({
        logUrl: this.logUrl,
        envelope: base64urlEncode(envelope),
        ...(integratedTime === undefined ? {} : { integratedTime }),
      }),
    });

    if (!response.ok) {
      throw new TypeError(`Sello log append failed with HTTP ${response.status}`);
    }

    return deserializeEntry(await response.json());
  }
}

export function serializeEntry(
  entry: TransparencyLogEntry,
): SerializedTransparencyLogEntry {
  return {
    logUrl: entry.logUrl,
    index: entry.index,
    integratedTime: entry.integratedTime,
    envelope: base64urlEncode(entry.envelope),
    proof: cloneJson(entry.proof),
  };
}

export function deserializeEntry(input: unknown): TransparencyLogEntry {
  if (!isRecord(input)) {
    throw new TypeError("Sello log entry must be an object");
  }

  const { logUrl, index, integratedTime, envelope, proof } = input;
  assertCanonicalLogUrl(logUrl, "entry.logUrl");

  if (!Number.isSafeInteger(index) || index < 0) {
    throw new TypeError("entry.index must be a non-negative safe integer");
  }

  if (
    typeof integratedTime !== "string" ||
    !/^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}Z$/.test(integratedTime)
  ) {
    throw new TypeError("entry.integratedTime must be an RFC 3339 UTC timestamp");
  }

  if (typeof envelope !== "string") {
    throw new TypeError("entry.envelope must be base64url");
  }

  return {
    logUrl,
    index,
    integratedTime,
    envelope: decodeBase64url(envelope, "entry.envelope"),
    proof: cloneJson(proof),
  };
}

export function toCanonicalLogUrl(url: string): CanonicalLogUrl {
  if (typeof url !== "string" || url.length === 0) {
    throw new TypeError("log URL must be a non-empty string");
  }

  const parsed = new URL(url);
  const path = parsed.pathname === "/" ? "/api" : parsed.pathname;
  const protocol =
    parsed.protocol === "http:" && isLocalHost(parsed.hostname) ? "https:" : parsed.protocol;
  const canonical = `${protocol}//${parsed.host}${path}`;

  assertCanonicalLogUrl(canonical, "logUrl");
  return canonical;
}

function normalizeEndpoint(url: string): string {
  const parsed = new URL(url);
  const path = parsed.pathname === "/" ? "/api" : parsed.pathname.replace(/\/$/, "");
  return `${parsed.protocol}//${parsed.host}${path}`;
}

function buildUrl(endpoint: string, path: string): string {
  return `${endpoint.replace(/\/$/, "")}${path}`;
}

function isLocalHost(hostname: string): boolean {
  return hostname === "localhost" || hostname === "127.0.0.1" || hostname === "::1";
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}

function cloneJson(value: unknown): unknown {
  if (value === null || typeof value !== "object") {
    return value;
  }

  return globalThis.structuredClone(value);
}

File: sdks/python/src/sello/logs.py
from __future__ import annotations

import json
from dataclasses import dataclass
from datetime import datetime, timezone
from typing import Any, Optional
from urllib.parse import urlparse
from urllib.request import Request, urlopen

from ._crypto import base64url_decode, base64url_encode


@dataclass(frozen=True)
class TransparencyLogEntry:
    log_url: str
    index: int
    integrated_time: str
    envelope: bytes
    proof: Any


class MemoryLog:
    def __init__(self, url: str):
        self.log_url = to_canonical_log_url(url)
        self.entries: list[TransparencyLogEntry] = []

    def append(self, envelope: bytes, integrated_time: Optional[str] = None) -> TransparencyLogEntry:
        entry = TransparencyLogEntry(
            log_url=self.log_url,
            index=len(self.entries),
            integrated_time=integrated_time or _now_utc_seconds(),
            envelope=bytes(envelope),
            proof={"type": "memory", "index": len(self.entries)},
        )
        self.entries.append(entry)
        return entry


class HttpLog:
    def __init__(self, url: str, *, endpoint: Optional[str] = None, headers: Optional[dict[str, str]] = None):
        self.log_url = to_canonical_log_url(url)
        self.endpoint = _normalize_endpoint(endpoint or url)
        self.headers = headers or {}

    def append(self, envelope: bytes, integrated_time: Optional[str] = None) -> TransparencyLogEntry:
        body = {
            "logUrl": self.log_url,
            "envelope": base64url_encode(bytes(envelope)),
        }
        if integrated_time is not None:
            body["integratedTime"] = integrated_time

        request = Request(
            _join_url(self.endpoint, "/entries"),
            data=json.dumps(body).encode("utf-8"),
            method="POST",
            headers={"content-type": "application/json", **self.headers},
        )
        with urlopen(request) as response:
            return deserialize_entry(json.loads(response.read().decode("utf-8")))


def memory(url: str) -> MemoryLog:
    return MemoryLog(url)


def http(url: str, *, endpoint: Optional[str] = None, headers: Optional[dict[str, str]] = None) -> HttpLog:
    return HttpLog(url, endpoint=endpoint, headers=headers)


def serialize_entry(entry: TransparencyLogEntry) -> dict[str, Any]:
    return {
        "logUrl": entry.log_url,
        "index": entry.index,
        "integratedTime": entry.integrated_time,
        "envelope": base64url_encode(entry.envelope),
        "proof": entry.proof,
    }


def deserialize_entry(value: dict[str, Any]) -> TransparencyLogEntry:
    return TransparencyLogEntry(
        log_url=to_canonical_log_url(value["logUrl"]),
        index=int(value["index"]),
        integrated_time=str(value["integratedTime"]),
        envelope=base64url_decode(str(value["envelope"]), "entry.envelope"),
        proof=value.get("proof"),
    )


def to_canonical_log_url(url: str) -> str:
    parsed = urlparse(url)
    if not parsed.scheme or not parsed.netloc:
        raise ValueError("log URL must be absolute")
    scheme = "https" if parsed.scheme == "http" and parsed.hostname in {"localhost", "127.0.0.1", "::1"} else parsed.scheme
    path = parsed.path if parsed.path and parsed.path != "/" else "/api"
    canonical = f"{scheme}://{parsed.netloc}{path}"
    if scheme != "https":
        raise ValueError("log URL must use https")
    if canonical.endswith("/") or parsed.query or parsed.fragment or parsed.username or parsed.password:
        raise ValueError("log URL must be canonical")
    return canonical


def _normalize_endpoint(url: str) -> str:
    parsed = urlparse(url)
    path = parsed.path if parsed.path and parsed.path != "/" else "/api"
    return f"{parsed.scheme}://{parsed.netloc}{path.rstrip('/')}"


def _join_url(endpoint: str, path: str) -> str:
    return f"{endpoint.rstrip('/')}{path}"


def _now_utc_seconds() -> str:
    return datetime.now(timezone.utc).replace(microsecond=0).isoformat().replace("+00:00", "Z")
